clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -520,7 +524,7 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                     // If idempotence is enabled only retry the request if the current producer id is the same as
                     // the producer id of the batch.
                     log.debug("Retrying batch to topic-partition {}. Sequence number : {}", batch.topicPartition,
-                            transactionManager.sequenceNumber(batch.topicPartition));
+                            batch.baseSequence());
                     reenqueueBatch(batch, now);
                 } else {
                     failBatch(batch, response, new OutOfOrderSequenceException("Attempted to retry sending a " +
```

Comment:
My understanding is that we reset the producerId if we find one partition for which we have expired all in-flight requests. My question is how this affects the in-flight requests of other partitions? If one of them happens to need retry, wouldn't we hit this case? But since the sequence number isn't changing, I think it should be safe to continue retrying with the old producerId.

Alternatively, maybe we should drain all partitions before resetting the producerId. That's a bit heavy-handed, but hopefully this case is rare in practice.

Reply:
This is a good point. Yes resetting the producer id due to one bad partition will affect in flight batches of other partitions which happen to need a retry. Our original take on this was 'bad luck, once you have expiring batches, this is just an additional bit of funkiness'. Personally, I don't think we need to over optimize this right now. If it becomes a usability issue, we can fix it in the future.

Reply:
I filed a JIRA to track this: https://issues.apache.org/jira/browse/KAFKA-5870

I think we should solve this --existing-- bug in 1.0.0, but I would like to do it after the code for KIP-192 goes in.

Reply:
Sounds good.